pkg/services/oauthserver/models.go

```diff
@@ -0,0 +1,77 @@
+package oauthserver
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/grafana/grafana/pkg/services/accesscontrol"
+	"gopkg.in/square/go-jose.v2"
+)
+
+const (
+	// TmpOrgID is the orgID we use while global service accounts are not supported.
+	TmpOrgID int64 = 1
+	// NoServiceAccountID is the ID we use for client that have no service account associated.
+	NoServiceAccountID int64 = 0
+
+	// List of scopes used to identify the impersonated user.
+	ScopeUsersSelf       = "users:self"
+	ScopeGlobalUsersSelf = "global.users:self"
+	ScopeTeamsSelf       = "teams:self"
+
+	// Supported encryptions
+	RS256 = "RS256"
+	ES256 = "ES256"
+)
+
+// OAuth2Server represents a service in charge of managing OAuth2 clients
+// and handling OAuth2 requests (token, introspection).
+type OAuth2Server interface {
+	// SaveExternalService creates or updates an external service in the database, it generates client_id and secrets and
+	// it ensures that the associated service account has the correct permissions.
+	SaveExternalService(ctx context.Context, cmd *ExternalServiceRegistration) (*ClientDTO, error)
```

Comment:
Should we use `Client` instead of `ExternalService` or vice versa, to only have one term?
All these functions we created have `ExternalService` but return a `Client` :sweat_smile:

Reply:
Discussed with @Jguer and @mgyongyosi, the suggestions we assessed were: `ExternalService` vs `Client` vs `App`. `ExternalService` or `App` were favored. `App` is brief and generic enough for us to use. We are going with `App`.